src/execution/mod.rs

```diff
@@ -68,36 +110,29 @@ impl Pump {
 /// whatever work it can.
 pub(crate) fn worker_main<A>(
     mut pumps_with_input_remaining: Vec<Pump>,
-    mut probes_with_timestamps_inflight: Vec<ProbeHandle<u64>>,
+    probe: ProbeHandle<u64>,
     interrupt_flag: &AtomicBool,
     worker: &mut timely::worker::Worker<A>,
 ) where
     A: timely::communication::Allocate,
 {
-    while (!pumps_with_input_remaining.is_empty() || !probes_with_timestamps_inflight.is_empty())
-        && !interrupt_flag.load(Ordering::Relaxed)
-    {
+    while !interrupt_flag.load(Ordering::Relaxed) && !probe.done() {
         if !pumps_with_input_remaining.is_empty() {
+            // We have input remaining, pump the pumps, and step the workers while
+            // there is work less than the current epoch to do.
             let mut updated_pumps = Vec::new();
             for mut pump in pumps_with_input_remaining.into_iter() {
                 pump.pump();
+                worker.step_while(|| probe.less_than(&pump.push_to_timely.time()));
```

Comment:
I'm really trying to wrap my head around this because I think it changes the behavior of our dataflows. Not that that's a bad thing! I just want to know for sure what behavior will manifest.

Doing a very careful read of [Timely's Monitoring Probes docs](https://timelydataflow.github.io/timely-dataflow/chapter_3/chapter_3_2.html): "Work until all data from before the input epoch has been captured."

This seems like the most intuitive behavior thinking of epochs as "batches" as long as there isn't any operators which "delay" in the dataflow (like anything that accumulated state but then emits at the oldest timestamp, not the latest). If there was, then this would cause hanging? Because then you'd need to keep introducing new data to get output, but this would prevent that. Perhaps this just generally a good constraint on our operators.

Reply:
Further head wrapping: does this make the dataflow... blocking? Non-concurrent? This is sort of saying that we won't have multiple epochs in flight inside the dataflow at the same time and I'm trying to think through what that means for resource utilization. Since every worker can do every task, maybe this isn't an issue; every worker churns through as much of an epoch as fast and thus if it finishes faster it'll get new input faster. But since progress tracking is across the whole cluster, imbalances in work could mean that some workers are waiting without work until a batch of slow keys on one worker is finished? I think that's less of a problem then the "each step is it's own compute" and provisioning for that. I don't think this is a reason to not make this change but it feels like there is another knob here we might need to eventually expose.